### ISSUE TITLE
ESLint: Enable the method-signature-style rule

### ISSUE THREAD

**Issue by acicovic** [Parsely/wp-parsely#2618]
## Is your feature request related to a problem?

The TL;DR is that we should enable the https://typescript-eslint.io/rules/method-signature-style rule in our code, and fix any violations.

To see why, this is an example (originally quoted by @sirreal, can be found [here](https://www.typescriptlang.org/play/?noUncheckedIndexedAccess=true&target=99&noImplicitOverride=true&exactOptionalPropertyTypes=true&noFallthroughCasesInSwitch=true&noPropertyAccessFromIndexSignature=true&ts=5.4.5#code/C4TwDgpgBAggdiAksCBbKBeKAlCBjAewCcATAHgGdgiBLOAcwBooBDBAPgG4AobulIgDMWeaAAUCVKBAAeKOCQqwEyNFADeUKi2ABXCgC4t1OvU5QAvr1CQoAMTIAVdpigAKGilRHHASkwuAEYEBAA2EGw8-BBCItCIDs4a7p5oPr5GwWERcOZWfHACwqJQAEIsFBBO0nIQCkrwSF4u6txQUDQUAKKhNPQ0geEeXumZIeGRbR3dvf2DEACMRolcU509fQPhAExGCU6r+YRwUoEVEEbllWSNqqguWK3t67NbFykjUBJU-hguqagAHTaPQUYHAFhEYAUADqngAFm4AOQQXoGJG+RhrGabeZLD5pL6SYC-f5ecE6fQUqGwhHI1E0dGY7EbOY7IzDQnfEkBDrkkFU7Q0uHAREotEYrEWIA)):

```
type AnyItem = Record<string, any>;

interface Post extends AnyItem { status: string; }

type F<T> = (item: T) => boolean;
interface IF<T> { (item: T): boolean; }

interface Base<T extends AnyItem> {
  isEligible(item: T): boolean;
  isEligible1: F<T>;
  isEligible2: IF<T>;
}

const base: Base<AnyItem> = {
  // Why is this is OK?
  isEligible: (item: Post) => item.status.startsWith('eli:'),

  // These are both type errors as expected.
  isEligible1: (item: Post) => item.status.startsWith('eli:'),
  isEligible2: (item: Post) => item.status.startsWith('eli:'),
}
```
Relatedly: https://x.com/mattpocockuk/status/1803753059480121593

**Comment by sirreal** [comment]
I endorse this change 😀
The next version of wp-scripts will enable this in the recommended ruleset.

**Comment by acicovic** [comment]
Note to our future selves: As we aren't currently using the last version of `wp-scripts` due to https://github.com/WordPress/gutenberg/issues/62202, we do want to add this rule, at least until we're able to upgrade.

**Comment by acicovic** [comment]
Related to #2624. If we become able to upgrade wp-scripts, then this rule will be auto-included. It probably makes sense to focus on #2624, as it should solve two issues at once.

**Comment by acicovic** [comment]
I finally went to fix this, since it was an easy win/change.
